### PR TITLE
UPC forward muon analysis

### PR DESCRIPTION
**Pull request by aga01935**
I added my code to analyze UPC muon in forward muon chamber. For now, this task access aod::BC to get CMUP11 trigger for run 246392
and use table aod::Muons to access the muon table

### CODE REVIEW

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    bool ispositive = kFALSE;
+    bool isnegative = kFALSE;
+    /*this code below is suggested by evgeny.
+    this code is now hardcoded for run 246392 as we are not sure if trigger id is same for all the runs*/
+    uint64_t classIndexMUP11 = 54; // 246392
+    bool isMUP11fired = bc.triggerMask() & (1ull << classIndexMUP11);
+    if (isMUP11fired) {
+      LOGF(info,
+           "mup11 fired");
+      hTriggerChecker->Fill(8);
```
Comment (ekryshen) [inline Analysis/Tasks/PWGUD/upcForward.cxx:87]: Why don't you use iSelectionCounter similar to other places?

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    int muontracknumber = 0;
+    TLorentzVector p1, p2, p;
+    bool ispositive = kFALSE;
+    bool isnegative = kFALSE;
+    /*this code below is suggested by evgeny.
+    this code is now hardcoded for run 246392 as we are not sure if trigger id is same for all the runs*/
+    uint64_t classIndexMUP11 = 54; // 246392
+    bool isMUP11fired = bc.triggerMask() & (1ull << classIndexMUP11);
+    if (isMUP11fired) {
```
Comment (ekryshen) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: I would just inserted if (!isMUP11fired) { return; }

File: Analysis/Tasks/PWGUD/upcForward.cxx (hunk 1 above)
Comment (ekryshen) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: May be remove hTriggerChecker histogram? It is a leftover from your trigger checks and all info should be available in hSelectionCounter

Comment (aga01935) [reply Analysis/Tasks/PWGUD/upcForward.cxx]: I made changes as suggested

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+#include "TLorentzVector.h"
+#include "AnalysisCore/TriggerAliases.h"
+using namespace std;
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+#define mmuon 0.1057 //mass of muon
+struct UPCForward {
+  //defining histograms
+  OutputObj<TH1D> hMass{TH1D(
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: Is there any particular reason for not having the histogram declaration on a single line?

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    ";#it{y},..;", 500, -10., 10.)};
+  OutputObj<TH1D> hCharge{TH1D(
+    "hCharge",
+    ";#it{charge},..;", 500, -10., 10.)};
+  OutputObj<TH1D> hSelectionCounter{TH1D(
+    "hSelectionCounter",
+    ";#it{Selection},..;", 30, 0., 30.)};
+  void process(aod::BC const& bc, aod::Muons const& tracksMuon)
+  {
+    int iSelectionCounter = 0;
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: if i understand the task correctly, the iSelectionCounter is not needed

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    hSelectionCounter->Fill(iSelectionCounter);
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: Here one can just say hSelectionCounter->Fill(0);

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    iSelectionCounter++;
+    hSelectionCounter->GetXaxis()->SetBinLabel(iSelectionCounter,
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: The bin labels for the histogram axis should be defined in the init function.

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    /*this code below is suggested by evgeny.
+    this code is now hardcoded for run 246392 as we are not sure if trigger id is same for all the runs*/
+    uint64_t classIndexMUP11 = 54; // 246392
+    bool isMUP11fired = bc.triggerMask() & (1ull << classIndexMUP11);
+    if (!isMUP11fired) {
+      return;
+    }
+    LOGF(info,
+         "mup11 fired");
+    hSelectionCounter->Fill(iSelectionCounter);
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: here you can just call hSelectionCounter->Fill(1);

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    iSelectionCounter++;
+    hSelectionCounter->GetXaxis()->SetBinLabel(iSelectionCounter,
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: same comment, axis labels should be set in init()

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+      }
+      if (muon.charge() < 0) {
+        p2.SetXYZM(muon.px(), muon.py(), muon.pz(), mmuon);
+        isnegative = kTRUE;
+      }
+    }
+    if (iMuonTracknumber != 2) {
+      return;
+    }
+    hSelectionCounter->Fill(iSelectionCounter);
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: same comment as before, iSelectionCounter is not needed

File: Analysis/Tasks/PWGUD/upcForward.cxx [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    iSelectionCounter++;
+    hSelectionCounter->GetXaxis()->SetBinLabel(iSelectionCounter,
```
Comment (iarsene) [inline Analysis/Tasks/PWGUD/upcForward.cxx]: bin labels to be defined in init(), see also another case below

**Review by iarsene — CHANGES_REQUESTED** [review]

File: Analysis/Tasks/PWGUD/upcForward.cxx (hunk 3 above)
Comment (aga01935) [reply Analysis/Tasks/PWGUD/upcForward.cxx]: It was formatted like this when I ran clang-format otherwise there is no particular reason for that.

Comment (aga01935) [reply Analysis/Tasks/PWGUD/upcForward.cxx]: I changed the format now.

File: Analysis/Tasks/PWGUD/upcForward.cxx (hunk 6 above)
Comment (aga01935) [reply Analysis/Tasks/PWGUD/upcForward.cxx]: I set the bin labels in init() function using

**Review by iarsene — APPROVED** [review]

**Pull request merged**